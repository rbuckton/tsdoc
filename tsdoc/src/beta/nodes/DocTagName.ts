import { SyntaxKind } from "./SyntaxKind";
import { SyntaxElement, ISyntaxParameters } from "./SyntaxElement";
import { ISyntaxElementSyntax } from "../syntax/ISyntaxElementSyntax";
import { DocTagNameSyntax } from "../syntax/tsdoc/elements/DocTagNameSyntax";

export interface IDocTagNameParameters extends ISyntaxParameters {
    text?: string;
}

export class DocTagName extends SyntaxElement {
    private _text: string | undefined;

    public constructor(parameters: IDocTagNameParameters = { }) {
        super(parameters);
        if (parameters.text !== undefined && !DocTagNameSyntax.isValidTagName(parameters.text)) {
            throw new Error('Invalid tag name.');
        }
        this._text = parameters.text;
    }

    /**
     * {@inheritdoc Node.kind}
     * @override
     */
    public get kind(): SyntaxKind.DocTagName {
        return SyntaxKind.DocTagName;
    }

    /**
     * {@inheritdoc Node.syntax}
     * @override
     */
    public get syntax(): ISyntaxElementSyntax<DocTagName> {
        return DocTagNameSyntax;
    }

    /**
     * Gets or sets the text of the tag name. Tag names must start with `@` followed by an ASCII letter and zero or more additional ASCII letters or decimal digits.
     */
    public get text(): string {
        return this._text || '@unknown';
    }

    public set text(value: string) {
        if (!DocTagNameSyntax.isValidTagName(value)) {
            throw new Error('Invalid tag name.');
        }
        if (this.text !== value) {
            this.beforeChange();
            this._text = value;
            this.afterChange();
        }
    }
}
