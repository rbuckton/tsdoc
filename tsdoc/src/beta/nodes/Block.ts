import { Content, IContentParameters } from "./Content";
import { Inline } from "./Inline";

export interface IBlockParameters extends IContentParameters {
}

export abstract class Block extends Content {
    // @ts-ignore
    private _blockBrand: never;

    public constructor(parameters?: IBlockParameters) {
        super(parameters);
    }

    /**
     * Gets the parent node of this node if that parent is a `Block`.
     */
    public get parentBlock(): Block | undefined {
        return this.parent && this.parent.isBlock() ? this.parent : undefined;
    }

    /**
     * Gets the previous sibling of this node if that sibling is a `Block`.
     */
    public get previousSiblingBlock(): Block | undefined {
        return this.previousSibling && this.previousSibling.isBlock() ? this.previousSibling : undefined;
    }

    /**
     * Gets the next sibling of this node if that sibling is a `Block`.
     */
    public get nextSiblingBlock(): Block | undefined {
        return this.nextSibling && this.nextSibling.isBlock() ? this.nextSibling : undefined;
    }

    /**
     * Gets the first of this node if that child is a `Block`.
     */
    public get firstChildBlock(): Block | undefined {
        return this.firstChild && this.firstChild.isBlock() ? this.firstChild : undefined;
    }

    /**
     * Gets the first of this node if that child is an `Inline`.
     */
    public get firstChildInline(): Inline | undefined {
        return this.firstChild && this.firstChild.isInline() ? this.firstChild : undefined;
    }

    /**
     * Gets the last of this node if that child is a `Block`.
     */
    public get lastChildBlock(): Block | undefined {
        return this.lastChild && this.lastChild.isBlock() ? this.lastChild : undefined;
    }

    /**
     * Gets the last of this node if that child is an `Inline`.
     */
    public get lastInlineChild(): Inline | undefined {
        return this.lastChild && this.lastChild.isInline() ? this.lastChild : undefined;
    }

    /**
     * @override
     */
    public isBlock(): true {
        return true;
    }
}
