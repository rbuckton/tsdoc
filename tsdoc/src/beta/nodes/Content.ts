import { INodeParameters, Node } from "./Node";
import { TSDocPrinter } from "../parser/TSDocPrinter";

export interface IContentParameters extends INodeParameters {
}

export abstract class Content extends Node {
    // @ts-ignore
    private _contentBrand: never;
    private _previousSibling: Content | undefined;
    private _nextSibling: Content | undefined;
    private _firstChild: Content | undefined;
    private _lastChild: Content | undefined;

    public constructor(parameters?: IContentParameters) {
        super(parameters);
    }

    /**
     * Gets the parent of this node if that parent is a `Content`.
     */
    public get parentContent(): Content | undefined {
        return this.parent && this.parent.isContent() ? this.parent : undefined;
    }

    /**
     * Gets the previous sibling of this node.
     */
    public get previousSibling(): Content | undefined {
        return this._previousSibling;
    }

    /**
     * Gets the next sibling of this node.
     */
    public get nextSibling(): Content | undefined {
        return this._nextSibling;
    }

    /**
     * Gets the first child of this node.
     */
    public get firstChild(): Content | undefined {
        return this._firstChild;
    }

    /**
     * Gets the last child of this node.
     */
    public get lastChild(): Content | undefined {
        return this._lastChild;
    }

    /** @override */
    public isContent(): true {
        return true;
    }

    /**
     * Appends a node as a child of this node.
     * @returns `true` if the child was appended; otherwise, `false`.
     */
    public appendChild(child: Content): boolean {
        if (this.canHaveChild(child) && child.canHaveParent(this) && child.isContent()) {
            child.removeNode();
            child.setParent(this);
            if (this._lastChild) {
                child._previousSibling = this._lastChild;
                this._lastChild = this._lastChild._nextSibling = child;
            } else {
                this._firstChild = this._lastChild = child;
            }
            return true;
        }
        return false;
    }

    /**
     * Removes a child node from this node.
     * @returns `true` if the child was removed; otherwise, `false`.
     */
    public removeChild(child: Content): boolean {
        if (child.parent === this && this.canHaveChild(child) && child.canHaveParent(this) && child.isContent()) {
            child.removeNode();
            return true;
        }
        return false;
    }

    /**
     * Inserts a new sibling node prior to this node.
     * @returns `true` if the sibling was inserted; otherwise, `false`.
     */
    public insertSiblingBefore(newSibling: Content): boolean {
        if (!this.parent ||
            !this.parent.isContent() ||
            !this.parent.canHaveChild(newSibling) ||
            !newSibling.isContent() ||
            !newSibling.canHaveParent(this.parent)) {
            return false;
        }
        newSibling.removeNode();
        newSibling._previousSibling = this._previousSibling;
        if (newSibling._previousSibling) {
            newSibling._previousSibling!._nextSibling = newSibling;
        }
        this._previousSibling = newSibling;
        newSibling._nextSibling = this;
        newSibling.setParent(this.parent);
        if (!newSibling._previousSibling) {
            this.parent._firstChild = newSibling;
        }
        return true;
    }

    /**
     * Inserts a new sibling node following this node.
     * @returns `true` if the sibling was inserted; otherwise, `false`.
     */
    public insertSiblingAfter(newSibling: Content): boolean {
        if (!this.parent ||
            !this.parent.isContent() ||
            !this.parent.canHaveChild(newSibling) ||
            !newSibling.isContent() ||
            !newSibling.canHaveParent(this.parent)) {
            return false;
        }
        newSibling.removeNode();
        newSibling._nextSibling = this._nextSibling;
        if (newSibling._nextSibling) {
            newSibling._nextSibling!._previousSibling = newSibling;
        }
        this._nextSibling = newSibling;
        newSibling._previousSibling = this;
        newSibling.setParent(this.parent);
        if (!newSibling._nextSibling) {
            this.parent._lastChild = newSibling;
        }
        return true;
    }

    /**
     * Replaces this node with the provided node in the document hierarchy.
     */
    public replaceNodeWith(newNode: Content): boolean {
        if (this.insertSiblingAfter(newNode)) {
            this.removeNode();
            return true;
        }
        return false;
    }

    /**
     * Removes this node from the document.
     * @override
     */
    public removeNode(): void {
        if (this.parent && this.parent.isContent()) {
            if (this._previousSibling) {
                this._previousSibling._nextSibling = this._nextSibling;
            } else if (this.parent && this.parent._firstChild === this) {
                this.parent._firstChild = this._nextSibling;
            }
            if (this._nextSibling) {
                this._nextSibling._previousSibling = this._previousSibling;
            } else if (this.parent && this.parent._lastChild === this) {
                this.parent._lastChild = this._previousSibling;
            }
        }
        this._previousSibling = undefined;
        this._nextSibling = undefined;
        this.setParent(undefined);
    }

    /**
     * Iterates through each syntactic and content element of this node.
     * @param cb The callback to execute for each node. If the callback returns a value other than `undefined`, iteration
     * stops and that value is returned.
     *
     * @override
     */
    public forEachNode<A extends any[], T>(cb: (node: Node, ...args: A) => T | undefined, ...args: A): T | undefined {
        return this.forEachSyntax(cb, ...args) || this.forEachChild(cb, ...args);
    }

    /**
     * Iterates through each content child of this node.
     * @param cb The callback to execute for each node. If the callback returns a value other than `undefined`, iteration
     * stops and that value is returned.
     */
    public forEachChild<A extends any[], T>(cb: (node: Content, ...args: A) => T | undefined, ...args: A): T | undefined {
        let child: Content | undefined = this._firstChild;
        while (child) {
            const next: Content | undefined = child._nextSibling;
            const result: T | undefined = cb(child, ...args);
            if (result !== undefined) {
                return result;
            }
            child = next;
        }
        return undefined;
    }

    protected abstract print(printer: TSDocPrinter): void;

    protected printChildren(printer: TSDocPrinter): void {
        this.forEachChild(child => child.print(printer));
    }
}