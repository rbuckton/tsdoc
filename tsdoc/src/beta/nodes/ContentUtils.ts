import { Content } from "./Content";
import { Inline, IInlineContainer, IInlineContainerParameters } from "./Inline";
import { Run } from "./Run";
import { Block, IBlockContainer, IBlockContainerParameters } from "./Block";
import { MarkdownParagraph } from "./MarkdownParagraph";
import { ListItemBase, IListItemContainer, IListItemContainerParameters } from "./ListItemBase";
import { ListMarker } from "./MarkdownListItem";
import { Token } from "../parser/Token";
import { MarkdownListItem } from "./MarkdownListItem";
import { MarkdownList } from "./MarkdownList";
import { ITableRowContainer, ITableRowContainerParameters, TableRowBase } from "./TableRowBase";
import { ITableCellContainer, ITableCellContainerParameters, TableCellBase } from "./TableCellBase";

export namespace ContentUtils {
    export function appendContent(node: Content & IBlockContainer, content: IBlockContainerParameters["content"]): boolean;
    export function appendContent(node: Content & IListItemContainer, content: IListItemContainerParameters["content"]): boolean;
    export function appendContent(node: Content & IInlineContainer, content: IInlineContainerParameters["content"]): boolean;
    export function appendContent(node: Content & ITableRowContainer, content: ITableRowContainerParameters["content"]): boolean;
    export function appendContent(node: Content & ITableCellContainer, content: ITableCellContainerParameters["content"]): boolean;
    export function appendContent(node: Content, content: string | Inline | Block | ListItemBase | TableRowBase | TableCellBase | ReadonlyArray<string | Inline | Block | ListItemBase | TableRowBase | TableCellBase> | undefined): boolean {
        return appendContentWorker(node, content);
    }

    function getDefaultListMarker(): ListMarker {
        return {
            ordered: false,
            bulletToken: Token.MinusToken,
            markerOffset: 0,
            padding: 2,
            tight: true
        };
    }

    function appendContentWorker(node: Content, content: string | Inline | Block | ListItemBase | ReadonlyArray<string | Inline | Block | ListItemBase> | undefined): boolean {
        if (content === undefined) {
            return false;
        }
        if (typeof content === 'string') {
            return appendContentWorker(node, new Run({ text: content }));
        }
        if ((Array.isArray as (array: any) => array is readonly unknown[])(content)) {
            let appendedAny: boolean = false;
            for (const item of content) {
                if (appendContentWorker(node, item)) {
                    appendedAny = true;
                }
            }
            return appendedAny;
        }
        for (;;) {
            if (node.appendChild(content)) {
                return true;
            }
            if (content.isInline()) {
                if (node.isBlockContainer() ||
                    node.isListItemContainer()) {
                    content = new MarkdownParagraph({ content });
                    continue;
                }
                if (node.isTableRowContainer() ||
                    node.isTableCellContainer()) {
                    throw new Error("Not yet implemented.");
                }
            }
            if (content.isListItem()) {
                if (node.isBlockContainer()) {
                    content = new MarkdownList({ content });
                    continue;
                }
            }
            if (node instanceof MarkdownList) {
                const listMarker: ListMarker = node.firstChildMarkdownListItem && node.firstChildMarkdownListItem.listMarker || getDefaultListMarker();
                content = new MarkdownListItem({ listMarker, content });
                continue;
            }
            if (node.isTableRowContainer() ||
                node.isTableCellContainer()) {
                throw new Error("Not yet implemented.");
            }
            return false;
        }
    }
}
