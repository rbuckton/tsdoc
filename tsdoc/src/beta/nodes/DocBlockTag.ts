import { SyntaxKind } from "./SyntaxKind";
import { DocTagName } from "./DocTagName";
import { Block, IBlockParameters } from "./Block";
import { Node } from "./Node";
import { Syntax } from "./Syntax";
import { TSDocPrinter } from "../parser/TSDocPrinter";

export interface IDocBlockTagParameters extends IBlockParameters {
    tagName?: DocTagName;
}

export class DocBlockTag extends Block {
    private _tagNameSyntax: DocTagName | undefined;

    public constructor(parameters?: IDocBlockTagParameters) {
        super(parameters);
        this.attachSyntax(this._tagNameSyntax = parameters && parameters.tagName);
    }

    /**
     * @override
     */
    public get kind(): SyntaxKind.DocBlockTag {
        return SyntaxKind.DocBlockTag;
    }

    public get tagNameSyntax(): DocTagName {
        if (!this._tagNameSyntax) {
            this.attachSyntax(this._tagNameSyntax = new DocTagName());
        }
        return this._tagNameSyntax;
    }

    public get tagName(): string {
        return this.tagNameSyntax.text;
    }

    public set tagName(value: string) {
        this.tagNameSyntax.text = value;
    }

    /**
     * @override
     */
    public isBlockContainer(): true {
        return true;
    }

    /**
     * @override
     */
    public canHaveParent(node: Node): boolean {
        return node.isDocument();
    }

    /**
     * @override
     */
    public getSyntax(): ReadonlyArray<Syntax> {
        const syntax: Syntax[] = [];
        if (this._tagNameSyntax) syntax.push(this._tagNameSyntax);
        return syntax;
    }

    /** @override */
    protected print(printer: TSDocPrinter): void {
        throw new Error("Not yet implemented");
    }
}
