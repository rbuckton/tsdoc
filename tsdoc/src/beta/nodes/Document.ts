import { SyntaxKind } from "./SyntaxKind";
import { Node, DocumentPosition } from "./Node";
import { Block, IBlockParameters } from "./Block";
import { MarkdownLinkReference } from "./MarkdownLinkReference";
import { MarkdownUtils } from "../parser/utils/MarkdownUtils";
import { MarkdownLinkLabel } from "./MarkdownLinkLabel";
import { TSDocPrinter } from "../parser/TSDocPrinter";

export interface IDocumentParameters extends IBlockParameters {
    text?: string;
}

export class Document extends Block {
    private _text: string | undefined;
    private _referenceMap = new Map<string, MarkdownLinkReference>();
    private _allReferences = new Map<string, MarkdownLinkReference[]>();

    public constructor(parameters?: IDocumentParameters) {
        super(parameters);
        this._text = parameters && parameters.text;
    }

    /** @override */
    public get kind(): SyntaxKind.Document {
        return SyntaxKind.Document;
    }

    public get text(): string | undefined {
        return this._text;
    }

    public get referenceMap(): ReadonlyMap<string, MarkdownLinkReference> {
        return this._referenceMap;
    }

    /** @override */
    public isBlockContainer(): true {
        return true;
    }

    /** @override */
    public isDocument(): true {
        return true;
    }

    /** @override */
    public canHaveParent(node: Node): false {
        return false;
    }

    /** @override */
    public canHaveChild(node: Node): boolean {
        return node.isBlock();
    }

    private static _compareReferences(x: MarkdownLinkReference, y: MarkdownLinkReference): number {
        switch (y.compareDocumentPosition(x)) {
            case DocumentPosition.Preceding: return -1;
            case DocumentPosition.Same: return 0;
            case DocumentPosition.Following: return 1;
            default: throw new Error("Illegal state.");
        }
    }

    private _attachLinkReference(node: MarkdownLinkReference): void {
        const refLabel: string = MarkdownUtils.normalizeLinkReference(node.label);
        if (!refLabel) return;

        let allReferences: MarkdownLinkReference[] | undefined = this._allReferences.get(refLabel);
        if (!allReferences) {
            allReferences = [];
            this._allReferences.set(refLabel, allReferences);
        }

        allReferences.push(node);
        allReferences.sort(Document._compareReferences);
        this._referenceMap.set(refLabel, allReferences[0]);
    }

    private _detachLinkReference(node: MarkdownLinkReference): void {
        const refLabel: string = MarkdownUtils.normalizeLinkReference(node.label);
        if (!refLabel) return;

        const allReferences: MarkdownLinkReference[] | undefined = this._allReferences.get(refLabel);
        if (!allReferences) return;

        const index: number = allReferences.indexOf(node);
        if (index < 0) return;

        allReferences.splice(index, 1);
        if (allReferences.length === 0) {
            this._referenceMap.delete(refLabel);
            this._allReferences.delete(refLabel);
        } else {
            allReferences.sort(Document._compareReferences);
            this._referenceMap.set(refLabel, allReferences[0]);
        }
    }

    /** @override */
    protected onNodeAttached(node: Node): void {
        if (node instanceof MarkdownLinkReference) {
            this._attachLinkReference(node);
        }
    }

    /** @override */
    protected onNodeDetached(node: Node): void {
        if (node instanceof MarkdownLinkReference) {
            this._detachLinkReference(node);
        }
    }

    /** @override */
    protected onNodeChanging(node: Node): void {
        if (node instanceof MarkdownLinkReference) {
            this._detachLinkReference(node);
        } else if (node instanceof MarkdownLinkLabel && node.parent instanceof MarkdownLinkReference) {
            this._detachLinkReference(node.parent);
        }
    }

    /** @override */
    protected onNodeChanged(node: Node): void {
        if (node instanceof MarkdownLinkReference) {
            this._attachLinkReference(node);
        } else if (node instanceof MarkdownLinkLabel && node.parent instanceof MarkdownLinkReference) {
            this._attachLinkReference(node.parent);
        }
    }

    /** @override */
    protected setOwnerDocument(ownerDocument: Document | undefined): void {
    }

    /** @override */
    protected print(printer: TSDocPrinter): void {
        this.printChildren(printer);
    }
}
