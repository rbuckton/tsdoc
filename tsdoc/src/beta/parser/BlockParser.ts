import { Scanner, IScannerState } from "./Scanner";
import { Token } from "./Token";
import { SyntaxKind } from "../nodes/SyntaxKind";
import { StartResult, ContinueResult, IBlockSyntaxParser } from "./blockParsers/IBlockSyntaxParser";
import { DocumentParser } from "./blockParsers/DocumentParser";
import { DocBlockTagParser } from "./blockParsers/DocBlockTagParser";
import { MarkdownBlockQuoteParser } from "./blockParsers/MarkdownBlockQuoteParser";
import { MarkdownHeadingParser } from "./blockParsers/MarkdownHeadingParser";
import { MarkdownCodeBlockParser } from "./blockParsers/MarkdownCodeBlockParser";
import { MarkdownHtmlBlockParser } from "./blockParsers/MarkdownHtmlBlockParser";
import { MarkdownThematicBreakParser } from "./blockParsers/MarkdownThematicBreakParser";
import { MarkdownListParser } from "./blockParsers/MarkdownListParser";
import { MarkdownListItemParser } from "./blockParsers/MarkdownListItemParser";
import { MarkdownParagraphParser } from "./blockParsers/MarkdownParagraphParser";
import { InlineParser } from "./InlineParser";
import { Document } from "../nodes/Document";
import { Node } from "../nodes/Node";
import { MarkdownCodeBlock } from "../nodes/MarkdownCodeBlock";
import { MarkdownParagraph } from "../nodes/MarkdownParagraph";
import { Block } from "../nodes/Block";
import { ContentWriter } from "./ContentWriter";
import { IMapping } from "./Preprocessor";
import { ParserBase } from "./ParserBase";

export class BlockParser extends ParserBase {
    private _line: number;
    private _linePos: number;
    private _indent: number;
    private _blank: boolean;
    private _indentStart: IScannerState | undefined;
    private _lastNextNonIndentPos: number;
    private _previousLineEnd: number;
    private _root!: Document;
    private _lastMatch: Block | undefined;
    private _hasUnmatchedBlocks: boolean;
    private _tip: Block | undefined;
    private _lastTip: Block | undefined;
    private _blockSyntaxParserMap: Map<SyntaxKind, IBlockSyntaxParser<Block>>;
    private _blockSyntaxParsers: ReadonlyArray<IBlockSyntaxParser<Block>>;

    public constructor(text: string, sourceMappings?: IMapping[]) {
        super(text, sourceMappings);
        this._line = 0;
        this._linePos = 0;
        this._indent = 0;
        this._blank = false;
        this._lastNextNonIndentPos= -1;
        this._previousLineEnd= -1;
        this._hasUnmatchedBlocks = false;
        this._blockSyntaxParsers = BlockParser.getDefaultBlockSyntaxParsers();
            this._blockSyntaxParserMap = new Map();
        for (const blockSyntaxParser of this._blockSyntaxParsers) {
            this._blockSyntaxParserMap.set(blockSyntaxParser.kind, blockSyntaxParser);
        }
    }

    /**
     * Gets the state of the scanner before the current indent.
     */
    public get indentStartState(): IScannerState | undefined {
        return this._indentStart;
    }

    /**
     * Gets the position at which the current indent started.
     */
    public get indentStartPos(): number {
        return this._indentStart ? this._indentStart.startPos : this._linePos;
    }

    /**
     * Gets the current indent depth relative to the current tip.
     */
    public get indent(): number {
        return this._indent;
    }

    /**
     * Gets the current tip.
     */
    public get tip(): Block | undefined {
        return this._tip;
    }

    /**
     * Gets a value indicating whether the current line is blank.
     */
    public get blank(): boolean {
        return this._blank;
    }

    /**
     * Gets the default parsers used to parse block syntax.
     */
    public static getDefaultBlockSyntaxParsers(): ReadonlyArray<IBlockSyntaxParser<Block>> {
        return [
            DocumentParser,
            DocBlockTagParser,
            MarkdownBlockQuoteParser,
            MarkdownHeadingParser,
            MarkdownCodeBlockParser,
            MarkdownHtmlBlockParser,
            MarkdownThematicBreakParser,
            MarkdownListParser,
            MarkdownListItemParser,
            MarkdownParagraphParser
        ];
    }

    public parse(): Document {
        this._line = 0;
        this._linePos = 0;
        this._indentStart = undefined;
        this._indent = 0;
        this._blank = false;
        this._lastNextNonIndentPos = -1;
        this._previousLineEnd = 0;
        this._root = new Document({ pos: 0, text: this.scanner.text });
        this._tip = this._root;
        this._lastMatch = this._root;
        this._hasUnmatchedBlocks = false;

        // prime the scanner with the first token
        this.scanner.scan();
        while (this.scanner.token() !== Token.EndOfFileToken) {
            this._parseLine();
            if (!Token.isLineEnding(this.scanner.token())) {
                throw new Error("Expected to be at end of line or end of input.");
            }
            this._previousLineEnd = this.scanner.startPos;
            this._linePos = this.scanner.pos;
            this._line++;
            this.scanner.scan();
        }
        while (this._tip) {
            this.finish(this._tip, this._previousLineEnd);
        }
        this._processInlines(this._root);
        return this._root;
    }

    private _parseLine(): void {
        let block: Block | undefined = this._root;
        this._lastTip = this._tip;

        // step 1: attempt to continue any open blocks that can accept the current line
        let nextBlock: Block | undefined = block.lastChildBlock;
        while (nextBlock && !this.getParserState(nextBlock).closed) {
            block = nextBlock;
            switch (this._continue(block)) {
                case ContinueResult.Matched:
                    // this.acceptIndent();
                    nextBlock = block.lastChildBlock;
                    continue;
                case ContinueResult.Finished:
                    return;
                case ContinueResult.Unmatched:
                    block = block.parentBlock || this._root;
                    break;
            }
            break;
        }

        this._hasUnmatchedBlocks = block !== this._lastTip;
        this._lastMatch = block;

        // step 2: if the line hasn't already been consumed, start a new block if possible
        let matchedLeaf: boolean = block.kind !== SyntaxKind.MarkdownParagraph && this._acceptsLines(block);
        while (block && !matchedLeaf) {
            switch (this._start(block)) {
                case StartResult.Container:
                    // this.acceptIndent();
                    block = this._tip;
                    continue;
                case StartResult.Leaf:
                    // this.acceptIndent();
                    block = this._tip;
                    matchedLeaf = true;
                    break;
            }
            break;
        }

        // step 3: any remaining content is inline content to be parsed later
        if (this._hasUnmatchedBlocks && !this._blank && this._tip && this._tip.kind === SyntaxKind.MarkdownParagraph) {
            // continue a paragraph
            this._acceptLine(this._tip);
        } else if (block) {
            this.finishUnmatchedBlocks();

            if (this._blank) {
                const lastChild: Block | undefined = block.lastChildBlock;
                if (lastChild) {
                    this.getParserState(lastChild).lastLineIsBlank = true;
                }
            }

            const lastLineIsBlank: boolean = this._blank &&
                !(block.kind === SyntaxKind.MarkdownBlockQuote ||
                    block.kind === SyntaxKind.MarkdownCodeBlock && !!(block as MarkdownCodeBlock).codeFence ||
                    block.kind === SyntaxKind.MarkdownListItem && !block.firstChild && this.lineMap.lineAt(block.pos) === this._line);

            let ancestor: Block | undefined = block;
            while (ancestor) {
                this.getParserState(ancestor).lastLineIsBlank = lastLineIsBlank;
                ancestor = ancestor.parentBlock;
            }

            this._acceptLine(block);
        } 

        // if for some reason we are not at the end of the line, skip to the end of the line
        this.scanner.scanLine();
    }

    private _acceptLine(block: Block): void {
        const blockSyntaxParser: IBlockSyntaxParser<Block> | undefined = this._blockSyntaxParserMap.get(block.kind);
        if (blockSyntaxParser && blockSyntaxParser.acceptLine) {
            blockSyntaxParser.acceptLine(this, block);
        } else if (!Token.isLineEnding(this.scanner.token()) && !this._blank) {
            const pos: number = this.indentStartPos;
            const block: MarkdownParagraph = this.pushBlock(new MarkdownParagraph({ pos, end: pos }));
            MarkdownParagraphParser.acceptLine(this, block);
        }
        this.scanner.scanLine();
    }

    private _start(container: Block): StartResult {
        this.nextNonIndent();
        for (const blockSyntaxParser of this._blockSyntaxParsers) {
            const startResult: StartResult = blockSyntaxParser.tryStart(this, container);
            if (startResult !== StartResult.Unmatched) {
                return startResult;
            }
        }
        return StartResult.Unmatched;
    }

    private _continue(block: Block): ContinueResult {
        this.nextNonIndent();
        const blockSyntaxParser: IBlockSyntaxParser<Block> | undefined = this._blockSyntaxParserMap.get(block.kind);
        return blockSyntaxParser ? blockSyntaxParser.tryContinue(this, block) : ContinueResult.Unmatched;
    }

    public finish(block: Block, end: number = this._previousLineEnd): void {
        const parent: Block | undefined = block.parentBlock;
        this.getParserState(block).closed = true;
        block.end = end;
        try {
            const blockSyntaxParser: IBlockSyntaxParser<Block> | undefined = this._blockSyntaxParserMap.get(block.kind);
            if (blockSyntaxParser) {
                blockSyntaxParser.finish(this, block);
            }
        } finally {
            this._tip = parent;
        }
    }

    public finishUnmatchedBlocks(): void {
        if (this._hasUnmatchedBlocks) {
            while (this._lastTip && this._lastTip !== this._lastMatch) {
                const parent: Block | undefined = this._lastTip.parentBlock;
                this.finish(this._lastTip);
                this._lastTip = parent;
            }
            this._hasUnmatchedBlocks = false;
        }
    }

    public pushBlock<T extends Block>(block: T, pos?: number, end?: number): T {
        if (pos !== undefined) {
            block.pos = pos;
            if (end !== undefined) {
                block.end = end;
            }
        }
        while (this._tip && !this._tip.canHaveChild(block)) {
            const parent: Block | undefined = this._tip.parentBlock;
            this.finish(this._tip);
            this._tip = parent;
        }
        if (this._tip) {
            this._tip.appendChild(block);
            this._tip = block;
        }
        return block;
    }

    public setTip<T extends Block>(block: T): T {
        this._tip = block;
        return block;
    }

    public nextNonIndent(): void {
        // check whether we've advanced the scanner.
        if (this._lastNextNonIndentPos === this.scanner.startPos) {
            return;
        }
        const column: number = this.scanner.startColumn;
        this._indentStart = this.scanner.getState();
        while (Token.isIndentCharacter(this.scanner.token())) {
            this.scanner.scan();
        }
        this._blank = Token.isLineEnding(this.scanner.token());
        this._indent = this.scanner.startColumn - column;
        this._lastNextNonIndentPos = this.scanner.startPos;
    }

    public acceptIndent(): void {
        this._blank = false;
        this._indentStart = undefined;
        this._indent = 0;
        this._lastNextNonIndentPos = -1;
    }

    public retreatToIndentStart(): void {
        this.scanner.setState(this._indentStart);
        this.acceptIndent();
    }

    public parseReferences(block: MarkdownParagraph): void {
        const content: ContentWriter | undefined = this.getParserState(block).content;
        if (content) {
            const parser: InlineParser = new InlineParser(this._root, content.toString(), content.mappings);
            for (const linkReference of parser.parseReferences()) {
                block.insertSiblingBefore(linkReference);
            }

            const scanner: Scanner = parser.scanner;
            if (scanner.token() === Token.EndOfFileToken) {
                this.getParserState(block).content = undefined;
                return;
            }

            if (scanner.startPos === 0) {
                return;
            }
            
            let sourcePos: number = scanner.startPos;
            const newContent: ContentWriter = new ContentWriter();
            newContent.addMapping(scanner.startPos);
            for (const sourceSegment of content.mappings) {
                if (sourceSegment.sourcePos > sourcePos) {
                    newContent.write(scanner.slice(sourcePos, sourceSegment.sourcePos));
                    newContent.addMapping(sourceSegment.sourcePos);
                    sourcePos = sourceSegment.sourcePos;
                }
            }
            newContent.write(scanner.slice(sourcePos));
            this.getParserState(block).content = newContent;
        }
    }

    private _acceptsLines(block: Node): boolean {
        const parser: IBlockSyntaxParser<Block> | undefined = this._blockSyntaxParserMap.get(block.kind);
        return parser && !!parser.acceptLine || false;
    }

    public endsWithBlankLine(block: Block): boolean {
        let current: Block | undefined = block;
        while (current) {
            if (this.getParserState(current).lastLineIsBlank) {
                return true;
            }
            if (!this.getParserState(current).lastLineChecked) {
                this.getParserState(current).lastLineChecked = true;
                if (current.isList()) {
                    current = current.lastChildListItem;
                    continue;
                }
                if (current.isListItem()) {
                    current = current.lastChildBlock;
                    continue;
                }
            }
            break;
        }
        return false;
    }

    private _processInlines(block: Block): void {
        if (block.isInlineContainer()) {
            const content: ContentWriter | undefined = this.getParserState(block).content;
            if (content && content.length) {
                const parser = new InlineParser(this._root, content.toString(), content.mappings);
                parser.parse(block);
            }
            this.getParserState(block).content = undefined;
        } else {
            block.forEachChild(child => {
                if (child.isBlock()) {
                    this._processInlines(child);
                }
            });
        }
    }
}
