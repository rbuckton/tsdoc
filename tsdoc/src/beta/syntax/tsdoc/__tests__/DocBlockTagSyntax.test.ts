// This file was autogenerated by ~/scripts/generateTests.js and should not be modified.

import { BlockParser } from "../../../parser/BlockParser";
import { Document } from "../../../nodes/Document";
import { SnapshotSerializer } from "../../../parser/__tests__/serializers";
import { TSDocConfiguration } from "../../../../configuration/TSDocConfiguration";

expect.addSnapshotSerializer(SnapshotSerializer.serializer);

function parse(text: string): Document {
    const parser: BlockParser = new BlockParser(new TSDocConfiguration(), text, undefined, false);
    return parser.parse();
}

it.each`
    tagName
    ${"@a"}
    ${"@ab"}
    ${"@a_"}
    ${"@a9"}
    ${"@a_9"}
`("DocBlockTag: $tagName", ({ tagName }: { tagName: string }) => {
    const document: Document = parse(tagName);
    expect(document).toMatchSnapshot();
});

