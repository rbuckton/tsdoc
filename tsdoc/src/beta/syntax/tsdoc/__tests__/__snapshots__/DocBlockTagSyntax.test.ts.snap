// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DocBlockTag: @a 1`] = `
Document {
  "kind": Document,
  "children": Array [
    TSDocBlockTag {
      "kind": TSDocBlockTag,
      "tagName": "@a",
      "pos": 0,
      "end": 2,
    },
  ],
  "pos": 0,
  "end": 2,
}
`;

exports[`DocBlockTag: @a_ 1`] = `
Document {
  "kind": Document,
  "children": Array [
    TSDocBlockTag {
      "kind": TSDocBlockTag,
      "tagName": "@a_",
      "pos": 0,
      "end": 3,
    },
  ],
  "pos": 0,
  "end": 3,
}
`;

exports[`DocBlockTag: @a_9 1`] = `
Document {
  "kind": Document,
  "children": Array [
    TSDocBlockTag {
      "kind": TSDocBlockTag,
      "tagName": "@a_9",
      "pos": 0,
      "end": 4,
    },
  ],
  "pos": 0,
  "end": 4,
}
`;

exports[`DocBlockTag: @a9 1`] = `
Document {
  "kind": Document,
  "children": Array [
    TSDocBlockTag {
      "kind": TSDocBlockTag,
      "tagName": "@a9",
      "pos": 0,
      "end": 3,
    },
  ],
  "pos": 0,
  "end": 3,
}
`;

exports[`DocBlockTag: @ab 1`] = `
Document {
  "kind": Document,
  "children": Array [
    TSDocBlockTag {
      "kind": TSDocBlockTag,
      "tagName": "@ab",
      "pos": 0,
      "end": 3,
    },
  ],
  "pos": 0,
  "end": 3,
}
`;
